nemo_aligner/algorithms/dpo.py

```diff
@@ -278,7 +279,7 @@ def save(self, extra_candidates=None, is_train_end=False):
         self.ckpt_callback.custom_save(monitor_candidates=monitor_candidates, is_train_end=is_train_end)
 
     def set_max_steps(self):
-        self.max_steps = self._train_dataloader_len + self.step
+        self.max_steps = (self._train_dataloader_len * (self.cfg.max_epochs - self.epoch)) + self.step
```

Comment:
Unfortunately for some reason beyond my comprehension, `max_epochs` must be set to `None` in PTL when unspecified, while `max_steps` must be set to `-1`. I don't think it makes sense but since NeMo users are likely to be familiar with PTL I'm afraid it's best to keep the same conventions (though I'm also fine with setting our own as long as we make them consistent in NeMo-Aligner).

Reply:
Having delved deeper into this, it appears that Aligner as a whole has no support for `max_epochs=null` and even the current main branch will break if you try to train with `max_epochs=null`. Adding support for that would require a bigger rewrite of our core training loop for every algorithm, because it's currently written with an outer for loop for epochs and an inner for loop for `max_steps` which not only breaks for `max_epochs=null` but it also breaks if max_steps is greater than a single epoch. Basically, we need an orthogonal PR with a large rewrite for how we account for max_steps being > single epoch, and also `max_epochs=null`. For now, I'm going to put an assert that max_epochs cannot be null.

Reply:
Ok sounds good, thanks for looking into it.

Reply:
this logic is quite complicated to get the max steps right. `max_epochs=null` can be easily supported with a default. but good to just leave it for now.

on max steps > single epoch:
* is this something we even want? the user first needs to know how many steps is in an epoch and then compute the max steps that way

Reply:
> on max steps > single epoch:
> 
> * is this something we even want? the user first needs to know how many steps is in an epoch and then compute the max steps that way

When `max_steps` is higher than the number of steps implicitly defined by the number of epochs, I think we should display a warning and use the latter (which is what I think we do right now -- probably minus the warning).

Someone who wants to use `max_steps` regardless of the value of `max_epochs` should set `max_epochs=null` (though of course we need to support that too :))

I'll create an issue regarding support for `max_epochs=null`.

Reply:
that sounds good! thanks

Reply:
> this logic is quite complicated to get the max steps right. `max_epochs=null` can be easily supported with a default. but good to just leave it for now.


I don't think you can add support with a default, that won't work. First, think about what `max_epochs=null` means, it means that max_steps is the main driver, and it solely controls for how long the trainer should run for. In the PTL case, max_steps can be longer than len(dataloader) or shorter than len(dataloader). But our code relies on an outer for loop which is `epoch_iter = range(self.epoch, self.cfg.max_epochs)` , which means you would still need need some concept of a max_epoch for it to work correctly. You can't just use a default.

What needs to happen is you need to calculate how many epochs for a given max_steps (when max_epochs=null) you need, and then round up. That would be the correct way to handle it.

Reply:
Possibly we can add the following logic to `set_max_steps`?

```
if self.cfg.max_epochs is None:
            if self.cfg.max_steps < 0:
                raise RuntimeError("max_steps cannot be -1 if max_epochs is null")
            max_epochs = math.ceil(self.cfg.max_steps / (self._train_dataloader_len + self.step))
            self.cfg.max_epochs = max_epochs
```

Reply:
Unless you force people to always set max_steps <= len(dataloader), in which case you can get away with using a default of max_epochs = 1, even when it's null, but then that sort of defeats the purpose of max_epochs=null.